# Experiment Naive Bayes NLP

In [1]:
import numpy as np
import pandas as pd

In [2]:
file_url = 'https://raw.githubusercontent.com/wongwara/Jobseeker_Baymax/d9d7fc4753ca407eba5423da6e1101e042b216a8/dataset/final_cleaned.csv'

In [3]:
df = pd.read_csv(file_url)

In [4]:
import warnings
warnings.filterwarnings("ignore")

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 638 entries, 0 to 637
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             638 non-null    int64  
 1   jobClassification      638 non-null    int64  
 2   state                  638 non-null    int64  
 3   teaser                 540 non-null    object 
 4   workType               638 non-null    int64  
 5   min_salary             638 non-null    int64  
 6   max_salary             638 non-null    int64  
 7   isRightToWorkRequired  638 non-null    int64  
 8   desktopAdTemplate      537 non-null    object 
 9   Python                 638 non-null    int64  
 10  SQL                    638 non-null    int64  
 11  R                      638 non-null    int64  
 12  Tableau                638 non-null    int64  
 13  SAS                    638 non-null    int64  
 14  Matlab                 638 non-null    int64  
 15  Hadoop

In [6]:
df.head()

Unnamed: 0  jobClassification  state  \
0           0                  0      1   
1           1                  1      1   
2           2                  1      1   
3           3                  2      0   
4           4                  3      0   

                                              teaser  workType  min_salary  \
0  Fantastic organisation seeks experienced Insig...         2       90000   
1  This role requires an individual with strong c...         2       90000   
2  One of Australia's leading financial service p...         2      110000   
3  Postdoctoral researcher in molecular evolution...         2       71509   
4  Postdoctoral researcher in molecular evolution...         2       71509   

   max_salary  isRightToWorkRequired  \
0      120000                      0   
1      110000                      0   
2      120000                      1   
3       90215                      0   
4       90215                      0   

                                   desktopAdTemplate  Python  ...  SAS  \
0  \n  \n    \n    \n      Insights Analyst – Onl...       0  ...    0   
1  \n  \n    \n    \n      Credit Risk Analyst \n...       0  ...    1   
2  \n  \n    \n    Data Analytics Recruitment Sol...       1  ...    1   
3  \n  \n    \n    \n      Postdoctoral Fellow \n...       0  ...    0   
4  \n  \n    \n    \n      Postdoctoral Fellow \n...       0  ...    0   

   Matlab  Hadoop  Spark  Java  Scala  recruiter  state_encoded  \
0       0       0      0     0      0          1              1   
1       0       0      0     0      0          1              1   
2       0       0      0     0      0          1              1   
3       0       0      0     1      0          0              0   
4       0       0      0     1      0          0              0   

         salary_section  salary_section_enc  
0  (100000.0, 110000.0]                 0.0  
1   (90000.0, 100000.0]                 1.0  
2  (110000.0, 120000.0]                 2.0  
3    (80000.0, 90000.0]                 3.0  
4    (80000.0, 90000.0]                 3.0  

[5 rows x 23 columns]

### Clean text

In [7]:
df['teaser']= df['teaser'].fillna('')

In [8]:
df['desktopAdTemplate']= df['desktopAdTemplate'].fillna('')

In [9]:
import pandas as pd
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import PorterStemmer
from sklearn.feature_extraction.text import TfidfVectorizer


# Clean the text data
df['teaser'] = df['teaser'].str.replace('[^\w\s]', '') # Remove punctuation
df['desktopAdTemplate'] = df['desktopAdTemplate'].str.replace('[^\w\s]', '') # Remove punctuation
df['teaser'] = df['teaser'].str.replace('\d+', '') # Remove digits
df['desktopAdTemplate'] = df['desktopAdTemplate'].str.replace('\d+', '') # Remove digits

# Normalize the text data
stop_words = set(stopwords.words('english'))
df['teaser'] = df['teaser'].apply(lambda x: ' '.join([word.lower() for word in x.split() if word.lower() not in stop_words]))
df['desktopAdTemplate'] = df['desktopAdTemplate'].apply(lambda x: ' '.join([word.lower() for word in x.split() if word.lower() not in stop_words]))

# Tokenize the text data
df['teaser'] = df['teaser'].apply(lambda x: word_tokenize(x))
df['desktopAdTemplate'] = df['desktopAdTemplate'].apply(lambda x: word_tokenize(x))

# Apply stemming
stemmer = PorterStemmer()
df['teaser'] = df['teaser'].apply(lambda x: [stemmer.stem(word) for word in x])
df['desktopAdTemplate'] = df['desktopAdTemplate'].apply(lambda x: [stemmer.stem(word) for word in x])

# # Create TF-IDF vectors
vectorizer = TfidfVectorizer()
teaser_tfidf = vectorizer.fit_transform(df['teaser'].apply(lambda x: ' '.join(x)))
desktopAdTemplate_tfidf = vectorizer.fit_transform(df['desktopAdTemplate'].apply(lambda x: ' '.join(x)))

# # Concatenate the TF-IDF vectors with the original dataframe
df = pd.concat([df.drop(['teaser', 'desktopAdTemplate'], axis=1), pd.DataFrame(teaser_tfidf.toarray()), pd.DataFrame(desktopAdTemplate_tfidf.toarray())], axis=1)

# Display the resulting dataframe
print(df.head())


   Unnamed: 0  jobClassification  state  workType  min_salary  max_salary  \
0           0                  0      1         2       90000      120000   
1           1                  1      1         2       90000      110000   
2           2                  1      1         2      110000      120000   
3           3                  2      0         2       71509       90215   
4           4                  3      0         2       71509       90215   

   isRightToWorkRequired  Python  SQL  R  ...  6803  6804  6805  6806  6807  \
0                      0       0    1  1  ...   0.0   0.0   0.0   0.0   0.0   
1                      0       0    1  1  ...   0.0   0.0   0.0   0.0   0.0   
2                      1       1    1  1  ...   0.0   0.0   0.0   0.0   0.0   
3                      0       0    0  0  ...   0.0   0.0   0.0   0.0   0.0   
4                      0       0    0  0  ...   0.0   0.0   0.0   0.0   0.0   

   6808  6809  6810  6811 6812  
0   0.0   0.0   0.0   0.0  0.

In [10]:
df.shape

(638, 7766)

In [11]:
df.columns

Index([           'Unnamed: 0',     'jobClassification',
                       'state',              'workType',
                  'min_salary',            'max_salary',
       'isRightToWorkRequired',                'Python',
                         'SQL',                     'R',
       ...
                          6803,                    6804,
                          6805,                    6806,
                          6807,                    6808,
                          6809,                    6810,
                          6811,                    6812],
      dtype='object', length=7766)

In [12]:
df = df.drop(['Unnamed: 0','state_encoded','min_salary','max_salary','salary_section'],axis=1)

In [13]:
df.columns

Index([    'jobClassification',                 'state',
                    'workType', 'isRightToWorkRequired',
                      'Python',                   'SQL',
                           'R',               'Tableau',
                         'SAS',                'Matlab',
       ...
                          6803,                    6804,
                          6805,                    6806,
                          6807,                    6808,
                          6809,                    6810,
                          6811,                    6812],
      dtype='object', length=7761)

In [14]:
df.shape

(638, 7761)

## Split data

In [15]:
# separate the target variable from the other features
y = df.pop('salary_section_enc')
X = df

from sklearn.model_selection import train_test_split
# split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [19]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(510, 7760)
(128, 7760)
(510,)
(128,)


In [20]:
# X_train.to_csv('/Users/thiradatiamklang/Desktop/NLP/AT2/cleandata/X_train.csv')
# X_test.to_csv('/Users/thiradatiamklang/Desktop/NLP/AT2/cleandata/X_test.csv')
# y_train.to_csv('/Users/thiradatiamklang/Desktop/NLP/AT2/cleandata/y_train.csv')
# y_test.to_csv('/Users/thiradatiamklang/Desktop/NLP/AT2/cleandata/y_test.csv')

## Naive Bayes

In [23]:
X_train.columns = X_train.columns.astype(str)
X_test.columns = X_test.columns.astype(str)
# Scale data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.fit_transform(X_test)

In [24]:
X_train

array([[-0.72618462, -0.67761149,  0.        , ..., -0.08889762,
        -0.04432422, -0.04432422],
       [-0.72618462, -0.67761149,  0.        , ..., -0.08889762,
        -0.04432422, -0.04432422],
       [ 0.10982928,  0.17991174,  0.        , ..., -0.08889762,
        -0.04432422, -0.04432422],
       ...,
       [ 0.52783623, -0.67761149,  0.        , ..., -0.08889762,
        -0.04432422, -0.04432422],
       [-0.72618462, -1.10637311,  0.        , ..., -0.08889762,
        -0.04432422, -0.04432422],
       [-0.72618462, -0.67761149,  0.        , ..., -0.08889762,
        -0.04432422, -0.04432422]])

In [25]:
# K-fold split
from sklearn.model_selection import StratifiedKFold, StratifiedShuffleSplit
cv = StratifiedShuffleSplit(n_splits=5, test_size=.20, random_state=42)

## 1. Naive Bayes with Grid search

In [26]:
params = {
    'var_smoothing': np.logspace(-10, -1, num=10)
}

In [27]:
from sklearn.naive_bayes import GaussianNB
gb = GaussianNB()

In [34]:
from sklearn.model_selection import GridSearchCV

grid_gb = GridSearchCV(gb, param_grid=params, scoring='accuracy', n_jobs =-1, cv=cv, verbose=1)

In [35]:
grid_gb.fit(X_train, y_train)

Fitting 5 folds for each of 10 candidates, totalling 50 fits


GridSearchCV(cv=StratifiedShuffleSplit(n_splits=5, random_state=42, test_size=0.2,
            train_size=None),
             estimator=GaussianNB(), n_jobs=-1,
             param_grid={'var_smoothing': array([1.e-10, 1.e-09, 1.e-08, 1.e-07, 1.e-06, 1.e-05, 1.e-04, 1.e-03,
       1.e-02, 1.e-01])},
             scoring='accuracy', verbose=1)

In [36]:
gb_grid_val_score = grid_gb.best_score_
print('Best Score:', gb_grid_val_score)
print('Best Estimator:', grid_gb.best_estimator_)

Best Score: 0.5078431372549019
Best Estimator: GaussianNB(var_smoothing=1e-10)


In [37]:
gb_grid= grid_gb.best_estimator_
y_pred = gb_grid.predict(X_test)

In [39]:
from sklearn.metrics import accuracy_score
gb_grid_score = accuracy_score(y_test, y_pred)
print('grid Model Accuracy:', gb_grid_score)


grid Model Accuracy: 0.5546875


## 2. Naive Bayes with Random search

In [41]:
from sklearn.model_selection import RandomizedSearchCV
rgrid_gb = RandomizedSearchCV(gb, param_distributions=params, scoring='accuracy', n_jobs =-1, cv=cv, verbose=1)

In [42]:
rgrid_gb.fit(X_train, y_train)

Fitting 5 folds for each of 10 candidates, totalling 50 fits


RandomizedSearchCV(cv=StratifiedShuffleSplit(n_splits=5, random_state=42, test_size=0.2,
            train_size=None),
                   estimator=GaussianNB(), n_jobs=-1,
                   param_distributions={'var_smoothing': array([1.e-10, 1.e-09, 1.e-08, 1.e-07, 1.e-06, 1.e-05, 1.e-04, 1.e-03,
       1.e-02, 1.e-01])},
                   scoring='accuracy', verbose=1)

In [43]:
gb_rgrid_val_score = rgrid_gb.best_score_
print('Best Score:', gb_rgrid_val_score)
print('Best Estimator:', rgrid_gb.best_estimator_)

Best Score: 0.5078431372549019
Best Estimator: GaussianNB(var_smoothing=1e-10)


In [52]:
gb_rgrid= rgrid_gb.best_estimator_
y_preds = gb_rgrid.predict(X_test)

In [53]:
gb_rgrid_score = accuracy_score(y_test, y_preds)
print('grid Model Accuracy:', gb_rgrid_score)

grid Model Accuracy: 0.5546875


## Result

Both grid search and random search provide the same best estimator which is _GaussianNB(var_smoothing=1e-10)_. The __best score of accuracy is 0.5078431372549019__, while the __test accuracy score is 0.5546875.__